mapie/calibrators/ccp/utils.py

```diff
@@ -0,0 +1,534 @@
+                    else:
+                        warn_ind[param] = [i]
+
+    if len(warn_ind) > 0:
+        warn_msg = ""
+        for param, inds in warn_ind.items():
+            warn_msg += (
+                f"The functions at index ({', '.join(map(str, inds))}) "
+                + "of the 'functions' argument, has an unknown optional "
+                + f"argument '{param}'.\n"
+            )
+        warnings.warn(
+            "WARNING: Unknown optional arguments.\n"
+            + warn_msg +
+            "The only recognized arguments are : 'X', 'y_pred' and 'z'. "
+            "The other optional arguments will act as parameters, "
+            "as it is always their default value which will be used."
+        )
+    if len(error_ind) > 0:
+        error_msg = ""
+        for param, inds in error_ind.items():
+            error_msg += (
+                f"The functions at index ({', '.join(map(str, inds))}) "
+                + "of the 'functions' argument, has an unknown required "
+                + f"argument '{param}'.\n"
+            )
+        raise ValueError(
+            "Forbidden required argument.\n"
+            f"{error_msg}"
+            "The only allowed required argument are : 'X', "
+            "'y_pred' and 'z'.\n"
+            "Note: You can use optional arguments if you want "
+            "to. They will act as parameters, as it is always "
+            "their default value which will be used."
+        )
+
+
+def sample_points(
+    X: ArrayLike,
+    points: Optional[Union[int, ArrayLike, Tuple[ArrayLike, ArrayLike]]]
+) -> NDArray:
+    """
+    Generate the ``points_`` attribute from the ``points`` and ``X`` arguments
+
+    Parameters
+    ----------
+    X : ArrayLike
+        Samples
+    points : Optional[Union[int, ArrayLike, Tuple[ArrayLike, ArrayLike]]]
+        If Array: List of data points, used as centers to compute
+        gaussian distances. Should be an array of shape (n_points, n_in).
+
+        If integer, the points will be sampled randomly from the ``X``
+        set, where ``X`` is the data give to the
+        ``GaussianCCP.fit`` method, which usually correspond to
+        the ``X`` argument of the ``MapieCCPRegressor.calibrate`` method
+        (unless you call ``GaussianCCP.fit(X)`` yourself).
+
+        You can pass a Tuple[ArrayLike, ArrayLike], to have a different
+        ``sigma`` value for each point. The two elements of the
+        tuple should be:
+         - Data points: 2D array of shape (n_points, n_in)
+         - Sigma values 2D array of shape (n_points, n_in) or (n_points, 1)
+        In this case, the ``sigma``, ``random_sigma`` and ``X`` argument are
+        ignored.
+
+        If ``None``, default to ``20``.
+
+    Returns
+    -------
+    NDArray
+        2D NDArray of points
+
+    Raises
+    ------
+    ValueError
+        If ``points`` is an invalid argument.
+    """
+    if points is None:
+        points = 20
```

Comment:
Why?

Reply:
Why not 20 points, in any case, the user could change it in input.

Reply:
It is not simple to find default values which work for all type/dimensions of datasets. Especially for the default standard deviation, it may definitely be improved with further experiments.

Reply:
Yeah, fair enough. No way to decide base on the dataset size?

Reply:
We could. It mostly depends on the computational power of the user, as too much points isn't really an issue. The biggest difficulty with the default parameters, is actually the standard deviation, as the optimal one not only depends on the dataset standard deviation, but also the number of points and the number of dimensions.